### DISCUSSION TITLE
getStaticProps triggered more often than `revalidate` time after 1st revalidate (ISR)

### DISCUSSION THREAD

**Discussion by LorrisSaintGenez** [vercel/next.js#44786] [Help]
Hi! 👋 

(I'm using next 12.3.4)

I'm having trouble understanding how revalidate actually works after the 1st revalidation.
From my understanding, if you set `revalidate` to `60`, `getStaticProps` will be triggered after 60sec on a page request:
- at build: getStaticProps is triggered
- before 60 seconds: getStaticProps is NOT triggered
- after 60 seconds : getStaticProps is triggered only for the first page request, and won't be called until 120 seconds (total, so 60 seconds after this trigger)
- and so on..

But once I'm deployed this simple code to Vercel:

```
import React from 'react';

const Index = () => <div>hi</div>;

const REVALIDATE_DELAY = 60;

export function getStaticProps() {
  // eslint-disable-next-line no-console
  console.log(
    `Hello there, are you really called every ${REVALIDATE_DELAY} seconds?`
  );

  return {
    props: {
      test: 'OK',
    },
    revalidate: REVALIDATE_DELAY,
  };
}

export default Index;
```

I can see clearly in the Functions Realtime Logs that the getStaticProps is called at **every page request**

<img width="1285" alt="Screenshot 2023-01-11 at 14 02 31" src="https://user-images.githubusercontent.com/6688748/211927386-ac388191-0f2a-45cc-a993-3a7dc0ea0e12.png">

Is there something I'm missing or misreading?

I've made a simple repository that reproduces this error: https://github.com/LorrisSaintGenez/revalidate-test

Thank you 🙏

**Comment by icyJoseph** [comment]
Do you observe the same locally? `yarn build && yarn serve`

When deployed on a cloud service, there might be several instances of your server, what's worse, each server might be spawning child tasks to run GetStaticProps, without blocking the main thread, not queueing event loop events.

That's how Next.js works on Vercel. GetStaticProps, GetServerSideProps and others are run as separate tasks from the core server. In addition to that you maybe hitting various servers.

Try to generate a unique identifier for each GetStaticProps run, and see if it happens twice.

**Reply by LorrisSaintGenez** [reply]
@icyJoseph I don't, it's only on Vercel.

I've tried, and after a few minutes, this is what I got:
<img width="1322" alt="Screenshot 2023-01-12 at 08 29 39" src="https://user-images.githubusercontent.com/6688748/212124752-c302a573-baa4-4ac9-860c-f835f3f3dbee.png">

Generated the ID this way
```
const REVALIDATE_DELAY = 30;

export function getStaticProps() {
  const id = crypto.randomBytes(16).toString('hex');

  console.log(id);

  return {
    props: {
      test: 'hi',
    },
    revalidate: REVALIDATE_DELAY,
  };
}
```

I'm a bit lost now, is that how it is supposed to work? 

Funny enough, it "works" as intended - meaning triggering the getStaticProps only once every 30sec max - when I update the UI from:
```
const Index = () => <div>hi</div>;
```
to
```
const Index = () => (
  <div>
    <h1>Hello!</h1>
    <p>It is currently --------------- {new Date().toString()}</p>
  </div>
);
```
([this commit](https://github.com/LorrisSaintGenez/revalidate-test/commit/b550249f827cf4d3e9eb5084c5053b0787646463))

**Reply by icyJoseph** [reply]
Almost, you should put the id outside. As long as it's not used in any client code the code elimination tool will rinse it from the client bundle. 

Now there'll be an id for each server that's built and we'll know if there's multiple running, or if it's the same instance running GSP several times.

I don't want to make more assumptions or pose any hypothesis, just yet anyway.

**Reply by LorrisSaintGenez** [reply]
It looks like it's the same instance running 🤔 
<img width="1340" alt="Screenshot 2023-01-12 at 13 57 53" src="https://user-images.githubusercontent.com/6688748/212189940-b8bf923d-2498-40f6-a81a-50e3bb6bbcde.png">

**Reply by icyJoseph** [reply]
Alright, let me start with a, that's weird...

As far as we know it is something triggered by the technique used to deploy Next.js apps in Vercel, a form of serverless deploy.

We know that it ultimately doesn't use the result of calling this function, because of your ` <p>It is currently --------------- {new Date().toString()}</p>` experiment. 

We also know that the GSP function is indeed, logging, so it must be executing the JavaScript code within it. 

So could we infer that the serverless deploy, activates the lambda that contains the GSP function for this page, runs it but doesn't update statically generated page unless a time has passed?

One last debug experiment:

```tsx
const Index = () => {
    if (typeof window === 'undefined') {
        console.log('Rendering  on Server');
    }
    return (
        <div>
            <h1>Hello!</h1>
            <p>It is currently --------------- {new Date().toString()}</p>
        </div>
    );
};
```

We can try to see if the page is also rendered, but ignored. Really strange stuff.

**Reply by LorrisSaintGenez** [reply]
So it is rendered but again, once every 30sec maximum - which is fine. With a project that simple, it's impossible it happened only to me 🤯 

<img width="1275" alt="Screenshot 2023-01-13 at 07 56 21" src="https://user-images.githubusercontent.com/6688748/212362860-9c892420-f4f2-4e61-aafc-bf93d69b9029.png">

**Reply by icyJoseph** [reply]
I know right? Maybe open a ticket with Vercel support, this feels like it is something specific to the way they deploy their apps. Mention also if you are using middleware or not, and specify as much as you can about your application.

**Reply by LorrisSaintGenez** [reply]
Yeah I did, thanks for the help anyway! I'll post an update if we find a solution
